### DISCUSSION TITLE
View type hinting

### DISCUSSION THREAD

**Discussion by ju5t** [laravel/framework#45763] [Q&A]
Our app has an action class that creates a pdf with Browsershot. Instead of testing the final PDF, we want to verify that the view we pass to Browsershot contains the data we expect. A simplified version of our class will look like this:

```php
class SomePdfAction
{
    // ...

    public function getView(): Illuminate\Contracts\View\Factory|Illuminate\Contracts\View\View
    {
        return view('some-pdf-view', ['param1' => $this-param]);
    }
}
```

As we return an object, we can't test it with `$this->view('some-pdf-view')`. As the helper expects a Blade-file as a sring, and not an object. Not a big deal as we can use `TestView` directly instead.

```php
$action = new SomePdfAction();
$view = new TestView($action->getView());
$view->assertSee('foo');
```

Our test passes, but PHPStorm warns that we're passing the wrong type to `TestView`. `TestView` expects a `View` and we're passing a contract. You would think this is an easy fix. Change the return type of `getView` in the action class to a `View` and you're done. Sadly, PHPStorm then warns you about returning the wrong type in `getView`, as the `view()` helper returns something different.

Is there a reason why there are differences here?

**Comment by henzeb** [comment]
Use docblocks
Put the getView statement in a variable and add a docblock with @var statement above it.

**Reply by ju5t** [reply]
Thanks for your reply, but it doesn't answer my question of why there are differences in the view types. You would imagine that the return type of `view()` is the same as what the construct of `TestView` expects.

**Reply by henzeb** [reply]
Well, probably because they wanted access to `gatherData` which is not on the View contract, nor on any of the other contracts the View implementation is implementing.  My guess is they didn't know they could use Closure::bind to access any parameters or methods inside the View implementation, or they considered this to much overhead for a unit test.

**Reply by ju5t** [reply]
That explains why `TestView` is more specific and requires a `View` object, I didn't know that. But what I don't understand is that `view()` has a docblock that says it returns the contract, while it actually does return the `View` object. Why that difference?

**Reply by henzeb** [reply]
They don't want you to know it's there, as it is not useful in normal situations. And they want to be flexible, hence the contract.
